File: virgin_python/Algorithms/binary_search.py
def binary_search(arr, target):

    l = 0
    r = len(arr) - 1

    # Keep shrinking until there are no values to check in each half
    while l <= r:

        mid = l + (r - l) // 2

        if arr[mid] == target:
            return mid
        
        # Shift the l to mid + 1 if the target is in the right half
        elif arr[mid] < target:
            l = mid + 1

        # Shift the r to mid - 1 if the target is in the left half
        else:
            r = mid - 1

    return None


# Time:
#   O(logn) => Number of steps until n becomes 1 is log_2(n)

# Space:
#   O(1) => No extra space needed

File: chad_c/Algorithms/binary_search.c
#include <stddef.h>

int binary_search(int *arr, size_t size, int target)
{
    if (size == 0)
    {
        return -1;
    }

    long l = 0;
    long r = (long)(size - 1);

    while (l <= r)
    {
        long mid = l + (r - l) / 2;

        if (arr[mid] == target)
        {
            return (int)mid;
        }

        else if (arr[mid] < target)
        {
            l = mid + 1;
        }

        else
        {
            r = mid - 1;
        }
    }

    return -1;
}
